charon/src/transform/monomorphize.rs

```diff
@@ -0,0 +1,543 @@
+        match self {
+            OptionHint::Some(_) => hint,
+            OptionHint::None => hint,
+            OptionHint::Hint(h) => h,
+        }
+    }
+}
+
+struct PassData {
+    // Map of (poly item, generic args) -> mono item
+    // None indicates the item hasn't been monomorphized yet
+    items: HashMap<(AnyTransId, GenericArgs), OptionHint<AnyTransId, (AnyTransId, GenericArgs)>>,
+    worklist: Vec<AnyTransId>,
+    visited: HashSet<AnyTransId>,
+}
+
+impl PassData {
+    fn new() -> Self {
+        PassData {
+            items: HashMap::new(),
+            worklist: Vec::new(),
+            visited: HashSet::new(),
+        }
+    }
+}
+
+impl TranslatedCrate {
+    fn find_trait_impl_and_gargs(self: &Self, kind: &TraitRefKind) -> (&TraitImpl, GenericArgs) {
+        match kind {
+            TraitRefKind::TraitImpl(impl_id, gargs) => {
+                let trait_impl = self.trait_impls.get(*impl_id).unwrap();
+                (trait_impl, gargs.clone())
+            }
+            TraitRefKind::ParentClause(p, _, clause) => {
+                let (trait_impl, _) = self.find_trait_impl_and_gargs(p);
+                let t_ref = trait_impl.parent_trait_refs.get(*clause).unwrap();
+                self.find_trait_impl_and_gargs(&t_ref.kind)
+            }
+            _ => panic!("Unexpected trait reference kind"),
+        }
+    }
+}
+
+#[derive(Visitor)]
+struct UsageVisitor<'a> {
+    data: &'a mut PassData,
+    krate: &'a TranslatedCrate,
+}
+impl UsageVisitor<'_> {
+    fn found_use(
+        &mut self,
+        id: &AnyTransId,
+        gargs: &GenericArgs,
+        default: OptionHint<AnyTransId, (AnyTransId, GenericArgs)>,
+    ) {
+        trace!("Mono: Found use: {:?} / {:?}", id, gargs);
+        self.data
+            .items
+            .entry((*id, gargs.clone()))
+            .or_insert(default);
+    }
+    fn found_use_ty(&mut self, id: &TypeDeclId, gargs: &GenericArgs) {
+        self.found_use(&AnyTransId::Type(*id), gargs, OptionHint::None);
+    }
+    fn found_use_fn(&mut self, id: &FunDeclId, gargs: &GenericArgs) {
+        self.found_use(&AnyTransId::Fun(*id), gargs, OptionHint::None);
+    }
+    fn found_use_fn_hinted(
+        &mut self,
+        id: &FunDeclId,
+        gargs: &GenericArgs,
+        (h_id, h_args): (FunDeclId, GenericArgs),
+    ) {
+        self.found_use(
+            &AnyTransId::Fun(*id),
+            gargs,
+            OptionHint::Hint((AnyTransId::Fun(h_id), h_args)),
+        );
+    }
+}
+impl VisitAst for UsageVisitor<'_> {
+    fn visit_aggregate_kind(&mut self, kind: &AggregateKind) -> ControlFlow<Infallible> {
+        match kind {
+            AggregateKind::Adt(TypeId::Adt(id), _, _, gargs) => self.found_use_ty(id, gargs),
+            AggregateKind::Closure(fun_id, gargs) => self.found_use_fn(fun_id, gargs),
+            _ => {}
+        }
+
+        Continue(())
+    }
+
+    fn visit_ty_kind(&mut self, kind: &TyKind) -> ControlFlow<Infallible> {
```

Comment:
Note: since you're recursing unconditionally, you might find it more convenient to use `enter_ty_kind` instead of `visit_ty_kind`. This is guaranteed to be called for all `TyKind`s without having to deal with recursive calls yourself.